effectsmanager.h

```diff
@@ -237,56 +237,45 @@ class EffectsManager : public IEffectsManager
 };
 
 // Define type aliases for effect (de)serialization functions for legibility reasons
-
 using EffectSerializer = function<void(nlohmann::json&, const ILEDEffect&)>;
 using EffectDeserializer = function<unique_ptr<ILEDEffect>(const nlohmann::json&)>;
 
-// Map with effect (de)serialization functions
-
-map<string, pair<EffectSerializer, EffectDeserializer>> gEffectSerializerFunctions;
-
-// Set the effect (de)serializer functions for one effect type
-
-template <typename T>
-void SetEffectFunctions()
+// Factory function to create a pair of effect (de)serialization functions for a given type
+template<typename T> 
+pair<EffectSerializer, EffectDeserializer> factoryPair() 
 {
-    EffectSerializer serializer = [](nlohmann::json& j, const ILEDEffect& effect) { to_json(j, dynamic_cast<const T&>(effect)); };
-    EffectDeserializer deserializer = [](const nlohmann::json& j) { return j.get<unique_ptr<T>>(); };
-
-    gEffectSerializerFunctions[T::EffectTypeName()] = make_pair(serializer, deserializer);
-}
+    EffectSerializer serializer = [](nlohmann::json& j, const ILEDEffect& effect) 
+    { 
+        to_json(j, dynamic_cast<const T&>(effect)); 
+    };
 
-// Register the (de)serializer functions for all effects we know about.
-// This function needs to be updated every time a new effect is added that needs to be (de)serialized.
+    EffectDeserializer deserializer = [](const nlohmann::json& j)
+    { 
+        return j.get<unique_ptr<T>>(); 
+    };
 
-inline void RegisterEffectSerializers()
-{
-    SetEffectFunctions<ColorWaveEffect>();
-    SetEffectFunctions<FireworksEffect>();
-    SetEffectFunctions<SolidColorFill>();
-    SetEffectFunctions<PaletteEffect>();
-    SetEffectFunctions<StarfieldEffect>();
-    SetEffectFunctions<MP4PlaybackEffect>();
+    return make_pair(serializer, deserializer);
 }
 
-//      Serialization Cheat Sheet
-//      --------------------------------------------------------------------------
-//      IController:      Serializes its properties and the canvas list
-//      ICanvas:          Serializes its properties including the IEffectManager
-//      IEffectMananger:  Serializes its ILEDEffects dynamically for each effect type
-//      ILEDEffect:       Serializes its properties
-//      ILEDFeature:      Serializes its properties
-//      ISocketChannel:   Serializes its properties
+// Map with effect (de)serialization functions
 
-// ILEDEffect --> JSON
+static const map<string, pair<EffectSerializer, EffectDeserializer>> to_from_json_map =
```

Comment:
I bet the macro cravings were strong right about here :-)

Reply:
...we...could...prevent...duplicating...the...typename...here...

...MUST...RESIST!!!!...

Reply:
We can, actually...

```c++
// Define type aliases for effect (de)serialization functions for legibility reasons
using EffectSerializer = function<void(nlohmann::json&, const ILEDEffect&)>;
using EffectDeserializer = function<unique_ptr<ILEDEffect>(const nlohmann::json&)>;

// Factory function to create a pair of effect (de)serialization functions for a given type
template<typename T> 
pair<string, pair<EffectSerializer, EffectDeserializer>> jsonPair() 
{
    EffectSerializer serializer = [](nlohmann::json& j, const ILEDEffect& effect) 
    { 
        to_json(j, dynamic_cast<const T&>(effect)); 
    };

    EffectDeserializer deserializer = [](const nlohmann::json& j)
    { 
        return j.get<unique_ptr<T>>(); 
    };

    return make_pair(typeid(T).name(), make_pair(serializer, deserializer));
}

// Map with effect (de)serialization functions

static const map<string, pair<EffectSerializer, EffectDeserializer>> to_from_json_map =
{
    jsonPair<ColorWaveEffect>(),
    jsonPair<FireworksEffect>(),
    jsonPair<SolidColorFill>(),
    jsonPair<PaletteEffect>(),
    jsonPair<StarfieldEffect>(),
    jsonPair<MP4PlaybackEffect>()
};
```

That'll be PR 21. I promise I'm going to leave this alone after that. ;)